Add BackendRegistry singleton class

ianthomas23 opened the pull request:
This is in preparation for #27663, moving the various hardcoded backend lists and dicts to a new `BackendRegistry` class that is intended to be the single source of truth for backends. There is no functional change.

There is a singleton `backend_registry` instance of the `BackendRegistry` class in `registry.py`. This is a similar implementation to that of the singleton `rcParams` instance of the `RcParams` class.

`BackendRegistry` includes the two public functions that are currently needed. There will eventually be a `list_all` function to list all backends including the built-in ones, those obtained via the `"module://blah.blah"` approach, and those registering themselves via entry points. The class internals will change too of course.

There are still some hardcoded backend names in `_fix_ipython_backend2gui`, but this will be addressed as part of the wider #27663 implementation.

I have not yet added an API change note. The `backend_registry` could be considered internal to Matplotlib except that IPython will also use it, so it could be thought of as fully public, or private with an exception for IPython. Other use of built-in backend lists by downstream libraries is probably restricted to use of `rcsetup.interactive_bk`, `non_interactive_bk` and `all_backends`. I have deprecated them in this PR to demonstrate that this is convoluted as they are module-level attributes that need to be deprecated using the recommendations of PEP 562. But probably I would be more inclined to keep them for the moment, until we are sure how this will all work in the end, and just pass through the `backend_registry` results.

(Edited to change name of singleton instance)

ianthomas23 commented on lib/matplotlib/backend_bases.py:
```diff
@@ -104,16 +104,12 @@ def _safe_pyplot_import():
+        from matplotlib.backends.registry import backendRegistry
+        backend = backendRegistry.framework_to_backend(current_framework)
+        if backend is None:
+            raise KeyError(backend)
```
This gives exactly the same error as prior to this PR. But probably it should be more informative.

timhoffm replied:
```suggestion
            raise KeyError(current_framework)
```
would reproduce the previous behavior. But I would go with
```suggestion
            raise RuntimeError(f"No suitable backend for the current GUI framework {current_framework!r}")
```

ianthomas23 commented on lib/matplotlib/rcsetup.py:
```diff
@@ -32,20 +33,34 @@
-                      'pdf', 'pgf', 'ps', 'svg', 'template']
-all_backends = interactive_bk + non_interactive_bk
+# Deprecation of module-level attributes using PEP 562
+_deprecated_interactive_bk = backendRegistry.list_builtin(BackendFilter.INTERACTIVE)
```
If we didn't deprecate these 3 attributes, this block of changes would be replaced by
```python
interactive_bk = backendRegistry.list_builtin(BackendFilter.INTERACTIVE)
non_interactive_bk = backendRegistry.list_builtin(BackendFilter.NON_INTERACTIVE)
all_backends = backendRegistry.list_builtin()
```

timhoffm commented on lib/matplotlib/backends/registry.py:
```diff
@@ -0,0 +1,61 @@
+            "headless": "agg",
+        }
+
+    def framework_to_backend(self, interactive_framework):
```
We can afford quite a long name for more clarity. 
```suggestion
    def gui_framework_to_backend(self, interactive_framework):
```
Or possibly even better
```suggestion
    def backend_for_gui_framework(self, interactive_framework):
```
You may then also shorten the parameter to just `framework` because *interactive* does not any information.

ianthomas23 replied:
I prefer the last suggestion.

timhoffm commented on lib/matplotlib/pyplot.py:
```diff
@@ -2486,7 +2481,7 @@ def polar(*args, **kwargs) -> list[Line2D]:
 if (rcParams["backend_fallback"]
         and rcParams._get_backend_or_none() in (  # type: ignore
-            set(rcsetup.interactive_bk) - {'WebAgg', 'nbAgg'})
+            set(backendRegistry.list_builtin(BackendFilter.INTERACTIVE_NON_WEB)))
```
```suggestion
            backendRegistry.list_builtin(BackendFilter.INTERACTIVE_NON_WEB))
```
The set is not needed anymore. it was only used to exclude WebAgg and nbAgg.

ianthomas23 replied:
Good catch.

ianthomas23 commented on lib/matplotlib/backends/registry.py:
```diff
@@ -0,0 +1,61 @@
+
+
+# Singleton
+backendRegistry = BackendRegistry()
```
Yes, I was following use of the existing singleton `rcParams = RcParams()`. But that is not a *good* reason! I'll change it.

ianthomas23 commented on lib/matplotlib/backends/registry.py:
```diff
@@ -0,0 +1,61 @@
+
+    This is the single source of truth for available backends.
+    """
+    def __init__(self):
```
Sure,

timhoffm reviewed (approved):
Looks good to me.

anntzer commented on lib/matplotlib/backends/registry.py:
```diff
@@ -0,0 +1,60 @@
+
+class BackendFilter(Enum):
+    INTERACTIVE = 0
+    INTERACTIVE_NON_WEB = 1
```
Why does ipython need to know about what backends are "web" or not?  Is that something that could be encoded in the required_interactive_framework attribute (e.g. by declaring a separate "web" value)?
This is mostly out of concern as to whether third-party backends ought to be able to declare themselves as "web" backends or not.

ianthomas23 replied:
This is nothing to do with IPython. This is a rearrangement of the existing code without any functional change. The 2 web-based backends were hardcoded in `pyplot.py` somewhere around line 2500, and that hard-coding has moved to this file.

anntzer commented on lib/matplotlib/pyplot.py:
```diff
@@ -2486,7 +2482,7 @@ def polar(*args, **kwargs) -> list[Line2D]:
 # is compatible with the current running interactive framework.
 if (rcParams["backend_fallback"]
         and rcParams._get_backend_or_none() in (  # type: ignore
-            set(rcsetup.interactive_bk) - {'WebAgg', 'nbAgg'})
```
(Replying to https://github.com/matplotlib/matplotlib/pull/27719#discussion_r1478201234.)  The logic here is actually wrong:
```sh
MATPLOTLIBRC=<(printf 'backend: qtcairo') python -c 'from PyQt6.QtWidgets import *; app = QApplication([]); from pylab import *; print(gcf().canvas)'
```
currently shows that the matplotlibrc entry gets ignored and the qtagg backend gets selected (because of this line).

I think the proper logic here should be "if backend_fallback is on and _get_backend_or_none() returns a non-None backend *that is not compatible with the running interactive framework*, then fallback to the autoselection logic".  If one fixes that, then there won't be any need to special-case the web backends anymore (and thus INTERACTIVE_NON_WEB can be removed from this PR) as they are compatible with any interactive framework.

We don't have to block this PR on fixing this (side) issue; I would suggest just keeping the special-casing for webagg/nbagg here for now and open a separate issue to track this.

ianthomas23 replied:
That all makes sense to me.

anntzer reviewed (changes requested):
I'll block based on https://github.com/matplotlib/matplotlib/pull/27719#discussion_r1478246139 though only as a note that INTERACTIVE_NON_WEB should be removed; anyone can dismiss once fixed.

ianthomas23 commented:
I've removed all of `INTERACTIVE_NON_WEB` and put the special case for webagg/nbagg back in `pyplot.py` as requested by @anntzer. Also rebased to pick up other fixes so hopefully all CI will pass.

QuLogic commented on lib/matplotlib/rcsetup.py:
```diff
@@ -32,20 +33,34 @@
+    if name in _deprecated_names_and_args:
+        arg = _deprecated_names_and_args[name]
+        _api.warn_deprecated(
+            "3.9.0",
```
We don't put micros here:
```suggestion
            "3.9",
```

ianthomas23 replied:
Done

QuLogic commented on lib/matplotlib/tests/test_backend_registry.py:
```diff
@@ -0,0 +1,56 @@
+    backends = backend_registry.list_builtin()
+    assert not has_duplicates(backends)
+    # Compare using sets as order is not important
+    assert set(backends) == set((
+        'GTK3Agg', 'GTK3Cairo', 'GTK4Agg', 'GTK4Cairo', 'MacOSX', 'nbAgg', 'QtAgg',
+        'QtCairo', 'Qt5Agg', 'Qt5Cairo', 'TkAgg', 'TkCairo', 'WebAgg', 'WX', 'WXAgg',
+        'WXCairo', 'agg', 'cairo', 'pdf', 'pgf', 'ps', 'svg', 'template',
+    ))
```
```suggestion
    assert {*backends} == {
        'GTK3Agg', 'GTK3Cairo', 'GTK4Agg', 'GTK4Cairo', 'MacOSX', 'nbAgg', 'QtAgg',
        'QtCairo', 'Qt5Agg', 'Qt5Cairo', 'TkAgg', 'TkCairo', 'WebAgg', 'WX', 'WXAgg',
        'WXCairo', 'agg', 'cairo', 'pdf', 'pgf', 'ps', 'svg', 'template',
    }
```

QuLogic commented on lib/matplotlib/tests/test_backend_registry.py:
```diff
@@ -0,0 +1,56 @@
+    backends = backend_registry.list_builtin(filter)
+    assert not has_duplicates(backends)
+    # Compare using sets as order is not important
+    assert set(backends) == set(expected)
```
```suggestion
    assert {*backends} == {*expected}
```

QuLogic commented on lib/matplotlib/rcsetup.py:
```diff
@@ -32,20 +33,34 @@
+}
+
+
+def __getattr__(name):
```
This should use `_api.caching_module_getattr` with suitable `_api.deprecated ` decorators.

ianthomas23 replied:
OK, I have switched to using `_api.caching_module_getattr`. I have kept the separate private attributes such as `_deprecated_interactive_bk` but these could just be moved to the individual deprecated properties if that is preferred.

QuLogic replied:
It doesn't look like they're used elsewhere, so they probably could be inlined. The lambdas could be changed to decorated methods if that helps readability.

ianthomas23 replied:
I have inlined them and changed to decorated methods as suggested, thanks.

QuLogic commented on lib/matplotlib/rcsetup.py:
```diff
@@ -32,20 +33,34 @@
+    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")
+
+
+def __dir__():
+    return sorted(globals().keys() | _deprecated_names_and_args.keys())
```
I'm not sure we usually attempt to preserve deprecated items in `__dir__`, unless this is something IPython needs?

ianthomas23 replied:
This was following the recommendations of PEP 562 (https://peps.python.org/pep-0562/). But now we are using the mpl-specific `_api.caching_module_getattr` approach rather than the PEP, I have removed this completely for consistency with the other similar uses.

ianthomas23 commented on lib/matplotlib/pyplot.py:
```diff
@@ -301,16 +302,11 @@ def switch_backend(newbackend: str) -> None:
+        if (current_framework and
+                (backend := backend_registry.backend_for_gui_framework(
+                current_framework))):
+            candidates = [backend]
```
Done

tacaswell commented:
Were we also going to add a way for third-party backends to registering themselves?

tacaswell commented:
> I very much think so. 😄 From the very top:

🐑 I failed my reading test for the day

ianthomas23 commented on lib/matplotlib/backends/registry.py:
```diff
@@ -0,0 +1,56 @@
+
+
+# Singleton
+backend_registry = BackendRegistry()
```
I lean towards the explicit and verbose approach, but maybe there is some middle ground. If I keep the files named as they are but import into `matplotlib.backends.__init__` using `from .registry import BackendFilter, backend_registry` then the standard usage becomes
```python
from matplotlib.backends import backend_registry
```
rather than
```python
from matplotlib.backends.registry import backend_registry
```
removing some verbosity.

This has the added advantage that you cannot import the `BackendRegistry` class in the same way which is good as nobody should need the class, only the singleton instance.

I have implemented this in commit 8f10a01c.

ianthomas23 commented:
> I very much think so. 😄 From the very top:
> 
> > This is in preparation for #27663, ([MNT]: Move Matplotlib backend mapping from IPython and support backends self-registering)

Yes, this PR is getting the ducks lined up ready for the real work in my next PR.

ianthomas23 commented:
The reduction is code coverage is due to changes I have had to make to `_safe_pyplot_import`. This function is not tested and is not used anywhere in the Matplotlib code base. It was added in #18316 (August 2020) and the one and only use of it was removed in #22925 (May 2022). So given that the function is private (leading underscore), not used and not tested, I think it should be removed.

QuLogic commented on lib/matplotlib/backends/registry.pyi:
```diff
@@ -0,0 +1,15 @@
+
+
+class BackendRegistry:
+    def __init__(self) -> None: ...
```
There's no `__init__` in the implementation; do we need  it in the typing?

ianthomas23 replied:
No we don't, I've removed it.

QuLogic commented on lib/matplotlib/rcsetup.pyi:
```diff
@@ -8,6 +8,8 @@ interactive_bk: list[str]
 non_interactive_bk: list[str]
 all_backends: list[str]
 
+def __dir__() -> list[str]: ...
```
This isn't added now.

ianthomas23 replied:
Done.
